src/Interpreters/DatabaseCatalog.cpp

```diff
@@ -338,9 +338,17 @@ DatabaseAndTable DatabaseCatalog::getTableImpl(
         database = it->second;
     }
 
-    auto table = database->tryGetTable(table_id.table_name, context_);
-    if (!table && exception)
-            exception->emplace(Exception(ErrorCodes::UNKNOWN_TABLE, "Table {} doesn't exist", table_id.getNameForLogs()));
+    StoragePtr table = nullptr;
+    try
+    {
+        table = database->getTable(table_id.table_name, context_);
```

Comment:
This has introduced lots of errors in system.errors when using table functions:

```
Row 1:
──────
server_shard:       1
version:            23.6.1.600
name:               UNKNOWN_TABLE
code:               60
quantity:           718
last_error_message: Table default.clusterAllReplicas doesn't exist
traceback:          DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)
DB::IDatabase::getTable(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::DatabaseCatalog::getTableImpl(DB::StorageID const&, std::__1::shared_ptr<DB::Context const>, std::__1::optional<DB::Exception>*) const
DB::DatabaseCatalog::tryGetTable(DB::StorageID const&, std::__1::shared_ptr<DB::Context const>) const
DB::Context::executeTableFunction(std::__1::shared_ptr<DB::IAST> const&, DB::ASTSelectQuery const*)
DB::getDatabaseAndTablesWithColumns(std::__1::vector<DB::ASTTableExpression const*, std::__1::allocator<DB::ASTTableExpression const*>> const&, std::__1::shared_ptr<DB::Context const>, bool, bool, bool)
DB::JoinedTables::resolveTables()
DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context> const&, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::PreparedSets>)
DB::createLocalPlan(std::__1::shared_ptr<DB::IAST> const&, DB::Block const&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long, unsigned long, unsigned long, std::__1::shared_ptr<DB::ParallelReplicasReadingCoordinator>, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>)
DB::ClusterProxy::SelectStreamFactory::createForShard(DB::Cluster::ShardInfo const&, std::__1::shared_ptr<DB::IAST> const&, DB::StorageID const&, std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>, std::__1::vector<std::__1::unique_ptr<DB::QueryPlan, std::__1::default_delete<DB::QueryPlan>>, std::__1::allocator<std::__1::unique_ptr<DB::QueryPlan, std::__1::default_delete<DB::QueryPlan>>>>&, std::__1::vector<DB::ClusterProxy::SelectStreamFactory::Shard, std::__1::allocator<DB::ClusterProxy::SelectStreamFactory::Shard>>&, unsigned int)
DB::ClusterProxy::executeQuery(DB::QueryPlan&, DB::Block const&, DB::QueryProcessingStage::Enum, DB::StorageID const&, std::__1::shared_ptr<DB::IAST> const&, DB::ClusterProxy::SelectStreamFactory&, Poco::Logger*, std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context const>, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::ExpressionActions> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Cluster> const&, std::__1::function<std::__1::shared_ptr<DB::IAST> (unsigned long)>)
DB::StorageDistributed::read(DB::QueryPlan&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageSnapshot> const&, DB::SelectQueryInfo&, std::__1::shared_ptr<DB::Context const>, DB::QueryProcessingStage::Enum, unsigned long, unsigned long)
DB::InterpreterSelectQuery::executeFetchColumns(DB::QueryProcessingStage::Enum, DB::QueryPlan&)
DB::InterpreterSelectQuery::executeImpl(DB::QueryPlan&, std::__1::optional<DB::Pipe>)
DB::InterpreterSelectQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::buildQueryPlan(DB::QueryPlan&)
DB::InterpreterSelectWithUnionQuery::execute()
DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum)
DB::TCPHandler::runImpl()
DB::TCPHandler::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)
```


Raising exceptions (with getTable) for things that aren't exceptions is bad since it gets reported and counted as an error, making monitoring impossible.

Reply:
Simplest reproducer:

```sql
production-01 :) select * from numbers(1);

SELECT *
FROM numbers(1)

Query id: 66f43d8d-e73a-4c14-a752-9be9d88a5640

┌─number─┐
│      0 │
└────────┘

1 row in set. Elapsed: 0.001 sec. 

production-01 :)     SELECT
        shardNum() as server_shard,
        version() as version,
        name,
        code,
        value AS quantity,
        last_error_message,
        arrayStringConcat(arrayMap(x -> demangle(addressToSymbol(x)), last_error_trace), '\n') AS traceback
    FROM system.errors WHERE name = 'UNKNOWN_TABLE'
    FORMAT Vertical
    SETTINGS allow_introspection_functions=1

SELECT
    shardNum() AS server_shard,
    version() AS version,
    name,
    code,
    value AS quantity,
    last_error_message,
    arrayStringConcat(arrayMap(x -> demangle(addressToSymbol(x)), last_error_trace), '\n') AS traceback
FROM system.errors
WHERE name = 'UNKNOWN_TABLE'
FORMAT Vertical
SETTINGS allow_introspection_functions = 1

Query id: 6d9c01a3-4cb1-4619-baba-26b45dc99c18

Row 1:
──────
server_shard:       0
version:            23.6.1.1
name:               UNKNOWN_TABLE
code:               60
quantity:           391
last_error_message: Table default.numbers doesn't exist
traceback:          Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, int)
DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool)
DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>(int, FormatStringHelperImpl<std::__1::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>::type, std::__1::type_identity<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>::type>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&&)
DB::IDatabase::getTable(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context const>) const
DB::DatabaseCatalog::getTableImpl(DB::StorageID const&, std::__1::shared_ptr<DB::Context const>, std::__1::optional<DB::Exception>*) const
DB::DatabaseCatalog::tryGetTable(DB::StorageID const&, std::__1::shared_ptr<DB::Context const>) const
DB::Context::executeTableFunction(std::__1::shared_ptr<DB::IAST> const&, DB::ASTSelectQuery const*)
DB::getDatabaseAndTablesWithColumns(std::__1::vector<DB::ASTTableExpression const*, std::__1::allocator<DB::ASTTableExpression const*>> const&, std::__1::shared_ptr<DB::Context const>, bool, bool, bool)
DB::JoinedTables::resolveTables()
DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context> const&, std::__1::optional<DB::Pipe>, std::__1::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::PreparedSets>)
DB::InterpreterSelectQuery::InterpreterSelectQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context> const&, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)
DB::InterpreterSelectWithUnionQuery::buildCurrentChildInterpreter(std::__1::shared_ptr<DB::IAST> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)
DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::__1::shared_ptr<DB::IAST> const&, std::__1::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)
std::__1::__unique_if<DB::InterpreterSelectWithUnionQuery>::__unique_single std::__1::make_unique[abi:v15000]<DB::InterpreterSelectWithUnionQuery, std::__1::shared_ptr<DB::IAST>&, std::__1::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&>(std::__1::shared_ptr<DB::IAST>&, std::__1::shared_ptr<DB::Context>&, DB::SelectQueryOptions const&)
DB::InterpreterFactory::get(std::__1::shared_ptr<DB::IAST>&, std::__1::shared_ptr<DB::Context>, DB::SelectQueryOptions const&)
DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*)
DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum)
DB::TCPHandler::runImpl()
DB::TCPHandler::run()
Poco::Net::TCPServerConnection::start()
Poco::Net::TCPServerDispatcher::run()
Poco::PooledThread::run()
Poco::ThreadImpl::runnableEntry(void*)


1 row in set. Elapsed: 0.002 sec.
```

I'll try reverting this part of the change (`tryGetTable` vs `getTable`) and see if we can add a test (although this is a global state, so it's hard to verify correctly)